energy-dao/src/external_sc_interactions/farm_interactions.rs

```diff
@@ -0,0 +1,252 @@
+
+use crate::{
+    common::errors::{
+        ERROR_BAD_PAYMENT_TOKENS, ERROR_EXTERNAL_CONTRACT_OUTPUT, ERROR_FARM_DOES_NOT_EXIST,
+        ERROR_UNBOND_TOO_SOON,
+    },
+    external_sc_interactions::farm_config::{UnstakeTokenAttributes, WrappedFarmTokenAttributes},
+};
+use common_structs::PaymentsVec;
+use locked_token_wrapper::wrapped_token;
+
+#[multiversx_sc::module]
+pub trait FarmInteractionsModule:
+    crate::external_sc_interactions::farm_config::FarmConfigModule
+    + crate::external_sc_interactions::farm_actions::FarmActionsModule
+    + crate::external_sc_interactions::locked_token_actions::LockedTokenModule
+    + crate::external_sc_interactions::fees_collector_interactions::FeesCollectorInteractionsModule
+    + lkmex_transfer::energy_transfer::EnergyTransferModule
+    + legacy_token_decode_module::LegacyTokenDecodeModule
+    + energy_query::EnergyQueryModule
+    + token_send::TokenSendModule
+    + utils::UtilsModule
+    + wrapped_token::WrappedTokenModule
+    + simple_lock::token_attributes::TokenAttributesModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+{
+    #[payable("*")]
+    #[endpoint(enterFarm)]
+    fn enter_farm_endpoint(&self, farm_address: ManagedAddress) -> EsdtTokenPayment {
+        let payment = self.call_value().single_esdt();
+        let mut farm_config_mapper = self.farm_config(&farm_address);
+        require!(!farm_config_mapper.is_empty(), ERROR_FARM_DOES_NOT_EXIST);
+        let farm_config = farm_config_mapper.get();
+        require!(
+            farm_config.farming_token_id == payment.token_identifier,
+            ERROR_BAD_PAYMENT_TOKENS
+        );
+        let mut enter_farm_payments = ManagedVec::new();
+        enter_farm_payments.push(payment);
+
+        let current_farm_position = EsdtTokenPayment::new(
+            farm_config.farm_token_id.clone(),
+            farm_config.farm_token_nonce,
+            farm_config.farm_staked_value.clone(),
+        );
+        let initial_total_farm_amount = current_farm_position.amount.clone();
+        if initial_total_farm_amount > 0 {
+            enter_farm_payments.push(current_farm_position);
+        }
+
+        let enter_farm_result = self.call_enter_farm(farm_address, enter_farm_payments);
+        let (new_farm_token, farm_rewards) = enter_farm_result.into_tuple();
+
+        require!(
+            new_farm_token.amount > initial_total_farm_amount,
+            ERROR_EXTERNAL_CONTRACT_OUTPUT
+        );
+
+        let user_farm_amount = &new_farm_token.amount - &initial_total_farm_amount;
+        self.update_farm_after_claim(
+            &farm_config,
+            &mut farm_config_mapper,
+            new_farm_token,
+            farm_rewards,
+        );
+
+        let caller = self.blockchain().get_caller();
+        let new_farm_config = farm_config_mapper.get();
+        let user_token_attributes = WrappedFarmTokenAttributes {
+            token_rps: new_farm_config.farm_rps,
```

Comment:
How is copying the farm rps in the attributes useful?

Reply:
Not sure I understand. It is the same logic as with the base farm tokens, that also contain the token `reward_per_share` attribute. I save it whenever `wrapped_farm_tokens` are minted, in order to know the moment when the user entered the contract, so I can later compute his rewards.